### PR TITLE
Server side pipe-ing 

### PR DESCRIPTION
**Pull request by MauritsMeijer**
Includes a test page named 'test'

### CODE REVIEW

**Comment by ogt** [comment]
Maurits, 
Github says that Failed — The Travis CI build failed.

File: .gitignore [hunk 1]
```diff
@@ -6,3 +6,4 @@
 !/.travis.yml
 /*.exe
 /*.project
+/node_modules
```
Comment (ogt) [inline .gitignore]: Node_modules should be part of the repository:
Take a look at http://www.futurealoof.com/posts/nodemodules-in-git.html

In summary:
- if you are building a module do not checkin node_modules
- If you are building an app do checkin node_modules

Comment (MauritsMeijer) [inline .gitignore]: Thanks, I think that was the problem. I'll look into the link you send
tomorrow.

File: index.js [hunk 2]
```diff
@@ -22,7 +27,19 @@ app.post('/*', function(req,res){
     cStream.on('error', function(err) {
         console.error(err);
     });
-    webCommand.webCommand(cmd,args, req, res, cStream);
+    if(req.query.pipes){
+    	var pipes= JSON.parse(req.query.pipes);
```
Comment (ogt) [inline index.js]: I want to use as much as the default HTTP protocol here - to minimize the things that I need to explain.

HTTP protocol defines a way to submit multi-value parameters - http://en.wikipedia.org/wiki/Query_string which is how we did the args.
In my view we should use the same protocol for the list of piped commands - the only difference is that unix command are now webcommands ie, urls of webservices that support our "webcommand protocol".

**Comment by ogt** [comment]
I notice that we use different indentations again.

Can you add in the Makefile test LINT-ing the index.js file (the way we did it in http://github.com/webcommand?).
You will also need to add the same .jshintrc file to the repo .

File: index.js [hunk 3]
```diff
@@ -22,7 +27,19 @@ app.post('/*', function(req,res){
...
+    	var curPipe=pipes.shift();
+    	var urlPipes='';
+    	if(pipes.length){
+    		urlPipes= '&pipes='+JSON.stringify(pipes);
+    	}
+    	var iStream= stream.through();
+    	iStream.pipe(request.post(curPipe+urlPipes)).pipe(res);
```
Comment (ogt) [inline index.js:48]: I think I mentioned this to you.. conacatenating `&pipes` will not work if there are no arguments. In that case you would need to add `?pipes`.
This means that you will need to extract from curPipes its args and if it exists do this vs that ... etc.

File: index.js (hunk 2 above)
Comment (MauritsMeijer) [inline index.js]: My two cents:
Without reading the rfc i am not sure the persistance of the ordering is ensured by the standard(seems to do it in node). And even though this is valid we replaced something equally valid with regards to the HTTP standard.

**Comment by MauritsMeijer** [comment]
.jshintrc was already in the repo. I told my editor to use spaces instead of taps. Please let me know if this solves the problem with indentation.

File: index.js (hunk 3 above)
Comment (MauritsMeijer) [inline index.js:48]: sharp! fixed it.

**Comment by ogt** [comment]
jshintrc was already in the repo but index.js was not "lint clean".
You can check that by doing `node_modules/.bin/jshint *.js`. (or if you haven't installed globally jshint via `npm install -g jshint` you can just do `jshint *.js`.

```
ithaca:webcommand-app odysseas$ jshint index.js 
index.js: line 33, col 4, Mixed spaces and tabs.
index.js: line 34, col 4, Mixed spaces and tabs.
index.js: line 35, col 4, Mixed spaces and tabs.
index.js: line 36, col 4, Mixed spaces and tabs.
index.js: line 37, col 4, Mixed spaces and tabs.
index.js: line 38, col 4, Mixed spaces and tabs.
index.js: line 38, col 13, 'curPipe' used out of scope.
index.js: line 39, col 4, Mixed spaces and tabs.
index.js: line 40, col 4, Mixed spaces and tabs.
index.js: line 41, col 4, Mixed spaces and tabs.
index.js: line 42, col 4, Mixed spaces and tabs.
index.js: line 43, col 4, Mixed spaces and tabs.
index.js: line 43, col 16, 'curPipe' used out of scope.
index.js: line 44, col 4, Mixed spaces and tabs.
index.js: line 44, col 93, Missing semicolon.
index.js: line 45, col 4, Mixed spaces and tabs.
index.js: line 46, col 4, Mixed spaces and tabs.
index.js: line 46, col 93, Missing semicolon.
index.js: line 47, col 4, Mixed spaces and tabs.
index.js: line 48, col 4, Mixed spaces and tabs.
index.js: line 49, col 4, Mixed spaces and tabs.
index.js: line 50, col 4, Mixed spaces and tabs.
index.js: line 50, col 35, 'curPipe' used out of scope.
index.js: line 51, col 4, Mixed spaces and tabs.
index.js: line 53, col 4, Mixed spaces and tabs.
index.js: line 36, col 24, 'curPipe' is defined but never used.

26 errors
```

Our `Makefile` contains the above command so when you do `make lint` it will try to do the same as above.
Linting is a dependency for testing so `make test` will also perform first a `make lint` and will fail if the repo isn' lint clean.
`make test` is set as the default way to test our repo via the configuration section

```
  "scripts": {
    "test": "make test"
  },
```

in `package.json` so `npm test` would do just `make test`.
This is what travis calls (`npm test`) so this is the reason that travis fails as well.
Finally when I meant to add a test file I meant to add an automated test in the test directory (make test will try to lint execute all the test files under the test directory - so not finding any there is also going to cause the test to fail.

Anyway, I will fix the jshint warnings - but I would leave it on you to add the automated test in the test directory.
Please do that before continuing with the other two issues that I have invited you to.

**Comment by ogt** [comment]
FYI, in unix you can do `expand -4` and this will find do the right thing... replacing any existing tab characters with the right number of spaces to maintain a indentation of 4 (it is smart and always works but you may want to do a diff afterwards... ) We should probably add `expand` in the webcommands :-)

**Comment by MauritsMeijer** [comment]
I think it's fixed now, no more lint errors. Eclipse lets you autoformat to the current indentation settings, so there is no rush to add expand to the webcommands:)

**Comment by ogt** [comment]
> I think it's fixed now, no more lint errors. Eclipse lets you autoformat
> to the current indentation settings, so there is no rush to add expand to
> the webcommands:)
> Yes - I fixed and push the fixes about 12 hrs ago

**Comment by ogt** [comment]
FYI, I wrapped the app functionality as a module - take a look at https://github.com/ogt/webcommand-express (the module) and how it gets used at https://github.com/ogt/clocserver .

**Pull request merged**